src/gridfs/upload.ts

```diff
@@ -147,10 +153,26 @@ export class GridFSBucketWriteStream extends Writable {
    * The stream is considered constructed when the indexes are done being created
    */
   override _construct(callback: (error?: Error | null) => void): void {
-    if (this.bucket.s.checkedIndexes) {
-      return process.nextTick(callback);
+    if (!this.bucket.s.calledOpenUploadStream) {
+      this.bucket.s.calledOpenUploadStream = true;
+
+      checkIndexes(this).then(
+        () => {
+          this.bucket.s.checkedIndexes = true;
+          this.bucket.emit('index');
+          callback();
+        },
+        error => {
+          if (error instanceof MongoOperationTimeoutError) {
+            return handleError(this, error, callback);
+          }
+          squashError(error);
+          callback();
+        }
+      );
+    } else {
+      callback();
```

Comment:
```suggestion
      return process.nextTick(callback);
```

In general, it's important to ensure that APIs are either 1. always async or 2. never async.  We aim to avoid "sometimes async" APIs.